core/sr-primitives/src/offence.rs

```diff
@@ -0,0 +1,128 @@
+/// that such timescale could be represented by a single `u128` value.
+pub type TimeSlot = u128;
+
+/// A trait implemented by an offence report.
+///
+/// This trait assumes that the offence is legitimate and was validated already.
+///
+/// Examples of offences include: a BABE equivocation or a GRANDPA unjustified vote.
+pub trait Offence<Offender> {
+	/// Identifier which is unique for this kind of an offence.
+	const ID: Kind;
+
+	/// The list of all offenders involved in this incident.
+	///
+	/// The list has no duplicates, so it is rather a set.
+	fn offenders(&self) -> Vec<Offender>;
+
+	/// The session index which is used for querying the validator set for the `slash_fraction`
+	/// function.
+	fn current_era_start_session_index(&self) -> u32; // TODO [slashing]: Should be a SessionIndex.
+
+	/// Return a validators count at the time when the offence took place.
+	fn validators_count(&self) -> u32;
+
+	/// A point in time when this offence happened.
+	///
+	/// The timescale is abstract and it doesn't have to be the same across different
+	/// implementations of this trait. For example, for GRANDPA it could represent a round number
+	/// and for BABE it could be a slot number.
+	fn time_slot(&self) -> TimeSlot;
+
+	/// A slash fraction of the total exposure that should be slashed for this
+	/// particular offence kind for the given parameters.
+	///
+	/// `offenders_count` - the count of unique offending authorities. It is >0.
+	/// `validators_count` - the cardinality of the validator set at the time of offence.
+	fn slash_fraction(
+		offenders_count: u32,
+		validators_count: u32,
+	) -> Perbill;
+}
+
+/// A trait for decoupling offence reporters from the actual handling of offence reports.
+pub trait ReportOffence<Reporter, Offender, O: Offence<Offender>> {
+	/// Report an `offence` and reward the `reporter`.
+	fn report_offence(reporter: Option<Reporter>, offence: O);
+}
+
+/// A trait to take action on an offence.
+///
+/// Used to decouple the module that handles offences and
+/// the one that should punish for those offences.
+pub trait OnOffenceHandler<Reporter, Offender> {
+	/// A handler for offence of particular kind.
+	///
+	/// Note that this contains a list of all previous offenders
+	/// as well. The implementer should cater for a case, where
+	/// the same authorities were reported for the same offence
+	/// in the past (see `OffenceCount`).
+	///
+	/// The vector of `slash_fraction` contains perbils
+	/// the authorities should be slashed and is computed
+	/// according to the `OffenceCount` already.
+	fn on_offence(
+		offenders: &[OffenceDetails<Reporter, Offender>],
+		slash_fraction: &[Perbill],
+	);
+}
+
+/// A details about an offending authority for a particular kind of offence.
+#[derive(Clone, PartialEq, Eq, Encode, Decode)]
+#[cfg_attr(feature = "std", derive(Debug))]
+pub struct OffenceDetails<Reporter, Offender> {
+	/// The offending authority id
+	pub offender: Offender,
```

Comment:
will this be the `IdentificationTuple` in practice?

Reply:
Yes.